python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -853,9 +854,15 @@ def _aggregate(
             else self._groupby.plc_groupby.aggregate(requests)
         )
 
-        for i, result in zip(column_included, results):
+        for i, result, val in zip(column_included, results, values):
             result_columns[i] = [
-                ColumnBase.from_pylibcudf(col) for col in result.columns()
+                ColumnBase.from_pylibcudf(col)._with_type_metadata(val.dtype)
+                if not isinstance(
+                    val.dtype,
+                    (DecimalDtype, CategoricalDtype),
+                )
```

Comment:
Shouldn't these be no-ops if they go through `_with_type_metadata`?

Reply:
This isn't a no-op for `decimal` and `categorical` types:
for `decimal`:
- The resulting `col` might be having a different precision than that of the original `val.dtype`, calling `_with_type_metadata` will corrupt it.
for `categorical`:
- Groupby aggregations on categorical columns some times result in different types, as need and expected for which we shouldn't be performing `_with_type_metadata`.

The right place to perform a `_with_type_metadata` is not here though. I have moved it to `agg`